### DISCUSSION TITLE
Swagger not handling null types

### DISCUSSION THREAD

**Discussion by to-bee** [fastapi/fastapi#10595] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
type: CategoryType | None = Query(default=None)
type: CategoryType | Query()
```

### Description

When definining a query parameter type of 'optional enum', then Swagger UI is not showing a dropdown item anymore. 
This did not happen with old fastapi version, using openapi 3.0.4. Current openapi version is: 3.1.0.
Is there a 'fastapi solution' to solve this problem?

Type: `category: CategoryType`
<img width="423" alt="image" src="https://github.com/tiangolo/fastapi/assets/6747788/ab4d84af-ff04-4c60-87f4-ee9ba3ca7aa5">

Type: `category: CategoryType | None`

<img width="525" alt="image" src="https://github.com/tiangolo/fastapi/assets/6747788/e007ccf2-de4d-4d75-97c6-e216559123b1">

Thanks for any advices or alterative approaches to solve this problem.

### Operating System

macOS

### FastAPI Version

0.103.0

### Pydantic Version

2.4.2

### Python Version

3.11.4

**Comment by Astrid153** [comment]
Downgrading pydantic to version 1.10.8 fixed the problem. But it's definitely not a solution.

**Reply by alex-pobeditel-2004** [reply]
As I understood, this problem will be fixed with https://github.com/tiangolo/fastapi/pull/9873 but it looks kind of stale

**Comment by YuriiMotov** [comment]
In current version (FastAPI 0.115.12, Pydantic 2.11.5) it looks different:
![Screenshot from 2025-05-30 08-11-45](https://github.com/user-attachments/assets/5081d142-af68-4d26-992c-90e5c957d387)
Now the dropdown list is shown, but the field is required (it's correct) and there is no possibility to send `null` using Swagger UI (this is not correct).

The parameter definition in openapi is:
```
        "parameters": [
          {
            "name": "cat_type",
            "in": "query",
            "required": true,
            "schema": {
              "anyOf": [
                {
                  "$ref": "#/components/schemas/CategoryType"
                },
                {
                  "type": "null"
                }
              ],
              "title": "Cat Type"
            }
          }
        ],
```

Adding `allowEmptyValue=True` would solve this problem:
```
        "parameters": [
          {
            "name": "cat_type",
            "in": "query",
            "required": true,
            "allowEmptyValue": true,
            "schema": {
              "anyOf": [
                {
                  "$ref": "#/components/schemas/CategoryType"
                },
                {
                  "type": "null"
                }
              ],
              "title": "Cat Type"
            }
          }
        ],
```
![Screenshot from 2025-05-30 08-39-04](https://github.com/user-attachments/assets/f22ef780-5b14-4a32-b400-5de42227f06f)

But I can't find way to add this `` in more elegant way than [by modifying openapi schema](https://github.com/fastapi/fastapi/discussions/11488?sort=new?sort=new#discussioncomment-9229143)

I think FastAPI (actually, Pydantic?) should add `"allowEmptyValue": true,` field for all nullable properties.

**Note:** there is one more problem here, in current version FastAPI treats empty parameter value as empty string instead of None. So, even sending request with "Send empty value" option will fail with validation error.
To solve this we have to use `BeforeValidator` to convert empty string to `None`: 
```python
@app.get("/")
async def index(
    cat_type: Annotated[
        CategoryType | None,
        BeforeValidator(lambda v: None if (v == "") else v),
    ],
):
    return cat_type
```
